src/test/app/AMM_test.cpp

```diff
@@ -0,0 +1,7610 @@
+        if (!swap)
+            return std::nullopt;
+        // optional transfer rate
+        auto const rate = getTransferRate(p);
+        if (!rate)
+            return std::nullopt;
+        auto const fee = getFee(p);
+        return {{pairs, *swap, *rate, fee}};
+    }
+
+    std::string
+    toString(STAmount const& a)
+    {
+        std::stringstream str;
+        str << a.getText() << "/" << to_string(a.issue().currency);
+        return str.str();
+    }
+
+    STAmount
+    mulratio(STAmount const& amt, std::uint32_t a, std::uint32_t b, bool round)
+    {
+        if (a == b)
+            return amt;
+        if (amt.native())
+            return toSTAmount(mulRatio(amt.xrp(), a, b, round), amt.issue());
+        return toSTAmount(mulRatio(amt.iou(), a, b, round), amt.issue());
+    }
+
+    void
+    swapOut(swapargs const& args)
+    {
+        auto const vp = std::get<steps>(args);
+        STAmount sout = std::get<STAmount>(args);
+        auto const fee = std::get<std::uint32_t>(args);
+        auto const rates = std::get<trates>(args);
+        STAmount resultOut = sout;
+        STAmount resultIn{};
+        STAmount sin{};
+        int limitingStep = vp.size();
+        STAmount limitStepOut{};
+        auto trate = [&](auto const& amt) {
+            auto const currency = to_string(amt.issue().currency);
+            return rates.find(currency) != rates.end() ? rates.at(currency)
+                                                       : QUALITY_ONE;
+        };
+        // swap out reverse
+        sin = sout;
+        for (auto it = vp.rbegin(); it != vp.rend(); ++it)
+        {
+            sout = mulratio(sin, trate(sin), QUALITY_ONE, true);
+            auto const [amts, amm] = *it;
+            // assume no amm limit
+            if (amm)
+            {
+                sin = swapAssetOut(amts, sout, fee);
+            }
+            else if (sout <= amts.out)
+            {
+                sin = Quality{amts}.ceil_out(amts, sout).in;
+            }
+            // limiting step
+            else
+            {
+                sin = amts.in;
+                limitingStep = vp.rend() - it - 1;
+                limitStepOut = amts.out;
+            }
+            resultIn = sin;
+        }
+        sin = limitStepOut;
+        // swap in if limiting step
+        for (int i = limitingStep + 1; i < vp.size(); ++i)
+        {
+            auto const [amts, amm] = vp[i];
+            sin = mulratio(sin, QUALITY_ONE, trate(sin), false);
+            if (amm)
+            {
+                sout = swapAssetIn(amts, sin, fee);
+            }
+            // assume there is no limiting step in fwd
+            else
+            {
+                sout = Quality{amts}.ceil_in(amts, sin).out;
+            }
+            sin = sout;
+            resultOut = sout;
+        }
+        std::cout << "in: " << toString(resultIn)
+                  << " out: " << toString(resultOut) << std::endl;
+    }
+
+    v
```

Comment:
I notice this test is consistently in the first few longest running tests. I suggest making it a high priority test so it can run sooner and get out of the way.
```
BEAST_DEFINE_TESTSUITE_PRIO(AMM, app, ripple, 1);
````

Reply:
Thanks, will add.